File: attempted/p165/p165.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <prime.h>

#define N 5000

typedef struct {
    long a, b;
} point_t;

typedef struct {
    point_t p0;
    point_t p1;
} segment_t;

#define NEXT(a) (((a)*(a)) % 50515093)

void generate(int num_segments, segment_t segments[])
{
    long s[] = {0,0,0,290797};

    int i;
    for(i = 0; i < num_segments; i++) {
        s[0] = NEXT(s[3]);
        s[1] = NEXT(s[0]);
        s[2] = NEXT(s[1]);
        s[3] = NEXT(s[2]);

        segments[i].p0.a = s[0] % 500;
        segments[i].p0.b = s[1] % 500;
        segments[i].p1.a = s[2] % 500;
        segments[i].p1.b = s[3] % 500;
    }
}

#define ABS(a) (((a) > 0) ? (a) : (-(a)))
bool check_st(long num, long den) {
    if(!num || !den)
        return false;

    if(num > 0 && den > 0) {
        return num < den;
    } else if((num < 0 && den > 0) || (num > 0 && den < 0)) {
        return false;
    } else {
        return num > den;
    }

    // not reachable
    return false;
}

bool check(segment_t *s0, segment_t *s1, segment_t *intersection)
{
    long a,b,c,d,e,f,g,h;
    a = s0->p0.a;
    b = s0->p0.b;
    c = s0->p1.a;
    d = s0->p1.b;
    e = s1->p0.a;
    f = s1->p0.b;
    g = s1->p1.a;
    h = s1->p1.b;

    long t_numerator = (c-g)*(f-h) - (d-h)*(e-g);
    long t_denominator = (b-d)*(e-g) - (a-c)*(f-h);

    if(!check_st(t_numerator, t_denominator))
        return false;

    long s_numerator = 0;
    long s_denominator = 0;
    if(e == g) {
        s_numerator = t_numerator * (b-d) + t_denominator*(d-h);
        s_denominator = t_denominator * (f-h);
    } else {
        s_numerator = t_numerator * (a-c) + t_denominator*(c-g);
        s_denominator = t_denominator * (e-g);
    }

    if(!check_st(s_numerator, s_denominator))
        return false;

    long x_numerator = ABS(t_numerator * a + (t_denominator - t_numerator)*c);
    long y_numerator = ABS(t_numerator * b + (t_denominator - t_numerator)*d);

    // encode rational (x, y) intersection as a pair of points
    intersection->p0.a = x_numerator;
    intersection->p0.b = ABS(t_denominator);
    intersection->p1.a = y_numerator;
    intersection->p1.b = ABS(t_denominator);

    return true;
}

inline bool same_fraction(point_t *a, point_t *b) {
    return b->b*a->a == a->b*b->a;
}

inline bool same_point(segment_t *a, segment_t *b) {
    return same_fraction(&(a->p0), &(b->p0)) && same_fraction(&(a->p1), &(b->p1));
}

int main(void) {
    segment_t *segments = malloc(N*sizeof(segment_t));

    int nalloc = 1 << 20;
    int n_intersections = 0;
    segment_t *intersections = malloc(nalloc*sizeof(segment_t));

    generate(N, segments);

    int i;
    for(i = 0; i < N; i++) {
        int j;
        for(j = 1 + i; j < N; j++) {
            bool hit = check(segments + i, segments + j, intersections + n_intersections);

            if(!hit)
                continue;

            n_intersections++;
            if(n_intersections == nalloc) {
                nalloc <<= 1;
                intersections = realloc(intersections, nalloc*sizeof(segment_t));
            }
        }
    }

    printf("%d total hits (including duplicates)\n", n_intersections);

    return 0;
    
    long np;
    long *p;
    primes(10000000, &p, &np);
    for(i = 0; i < n_intersections; i++) {
        long nx, dx, ny, dy;

        reduce(intersections[i].p0.a, intersections[i].p0.b, np, p, &nx, &dx);
        //printf("%ld/%ld => %ld/%ld\n", intersections[i].p0.a, intersections[i].p0.b, nx, dx);

        intersections[i].p0.a = nx;
        intersections[i].p0.b = dx;

        reduce(intersections[i].p1.a, intersections[i].p1.b, np, p, &ny, &dy);
        //printf("%ld/%ld => %ld/%ld\n", intersections[i].p1.a, intersections[i].p1.b, ny, dy);
        intersections[i].p1.a = ny;
        intersections[i].p1.b = dy;

        //printf("%ld/%ld\t%ld/%ld\n", nx, dx, ny, dy);
    }

    /*
    int n_unique = 0;
    for(i = 0; i < n_intersections; i++) {
        if(!(i % 100))
            printf("working on item %d\n", i);
        n_unique++;
        int j = 1 + i;
        while(j < n_intersections) {
            if(same_point(intersections + i, intersections + j)) {
                intersections[j] = intersections[--n_intersections];
            }
            j++;
        }
    }

    printf("%d unique hits\n", n_unique);
    */

    return 0;
}

File: attempted/p165/p165.py
#!/usr/bin/python

from fraction import Fraction

class LineSegment:
    def __init__(self, p0, p1):
        (a, b) = p0
        (c, d) = p1

        self.a = a
        self.b = b
        self.c = c
        self.d = d

def check(l0, l1):
    a = l0.a
    b = l0.b
    c = l0.c
    d = l0.d

    e = l1.a
    f = l1.b
    g = l1.c
    h = l1.d

    num = (c-g)*(f-h) - (d-h)*(e-g)
    den = (b-d)*(e-g) - (a-c)*(f-h)

    if den == 0:
        return None # no intersection points

    t = float(num) / den

    #print('t == %s; tf == %s' % (str(t), str(tf)))

    if 0 < t < 1:
        if e == g:
            s = (t * (b-d) + (d-h)) / (f-h)
        else:
            s = (t * (a-c) + (c-g)) / (e-g)

        #print('s == %s; sf == %s' % (str(s), str(sf)))

        if 0 < s < 1:
            tf = Fraction(num, den)
            if e == g:
                sf = (tf * (b-d) + (d-h)) * Fraction(1, f-h)
            else:
                sf = (tf * (a-c) + (c-g)) * Fraction(1, e-g)

            #if 0 < sf < 1 and 0 < tf < 1:
            # compute intersection point and return
            #x = t*a + (1-t)*c
            #y = t*b + (1-t)*d 
            #return (x, y)

            xf = tf*a + (1-tf)*c
            yf = tf*b + (1-tf)*d 
            return (xf, yf)

    return None

def generate(N):
    items = []
    s4 = 290797

    for i in range(N):
        s1 = (s4*s4) % 50515093
        s2 = (s1*s1) % 50515093
        s3 = (s2*s2) % 50515093
        s4 = (s3*s3) % 50515093

        if i == 0:
            print('(%d, %d), (%d, %d)' % (s1 % 500, s2 % 500, s3 % 500, s4 % 500))

        segment = LineSegment((s1 % 500, s2 % 500), (s3 % 500, s4 % 500))
        items.append(segment)

    return items

segments = generate(5000)

p1 = LineSegment((27, 44), (12, 32))
p2 = LineSegment((46, 53), (17, 62))
p3 = LineSegment((46, 70), (22, 40))

print('check p1, p2: ' + str(check(p1, p2)))
print('check p2, p3: ' + str(check(p2, p3)))
print('check p1, p3: ' + str(check(p1, p3)))

intersections = set([])
num_raw = 0
for i in range(len(segments)):
    #if i % 100 == 0:
    #    print('working on segment %d' % i)
    a = segments[i]
    for j in range(1 + i, len(segments)):
        b = segments[j]

        x = check(a, b)

        if x is not None:
            (xf, yf) = x
            num_raw += 1
            print('%s\t%s' % (str(xf), str(yf)))
            intersections.add((xf.tuple(), yf.tuple()))

print('total number of intersections: %d' % num_raw)
print('number of unique intersections: %d' % len(intersections))

